numpyro/primitives.py

```diff
@@ -80,7 +80,31 @@ def __call__(self, *args, **kwargs):
             return self.fn(*args, **kwargs)
 
 
-def sample(name, fn, obs=None, rng_key=None, sample_shape=(), infer=None):
+def _masked_observe(name, fn, obs, obs_mask, **kwargs):
+    # Split into two auxiliary sample sites.
+    with numpyro.handlers.mask(mask=obs_mask):
+        observed = sample(f"{name}_observed", fn, **kwargs, obs=obs)
+    with numpyro.handlers.mask(mask=(not obs_mask if isinstance(obs_mask, bool) else ~obs_mask)):
+        unobserved = sample(f"{name}_unobserved", fn, **kwargs)
+
+    # Interleave observed and unobserved events.
+    shape = jnp.shape(obs_mask) + (1,) * fn.event_dim
+    batch_mask = jnp.reshape(obs_mask, shape)
+    try:
+        value = jnp.where(batch_mask, observed, unobserved)
+    except RuntimeError as e:
+        if "must match the size of tensor" in str(e):
```

Comment:
Could you port Pyro's [death test](https://github.com/pyro-ppl/pyro/pull/2772/files#diff-84b43e1d1f62fa4c1d721f194efd2ffe0dcea3e07522cc505109a401af596cd9R2266) to NumPyro? I suspect this error string is torch-specific, and you'll need a different error string and possibly exception type for JAX.

Alternatively, you could just omit this fancy error handling for now.

Reply:
Thanks, I'll omit this for simplicity. :)